src/poetry/packages/locker.py

```diff
@@ -204,102 +205,104 @@ def locked_repository(self, with_dev_reqs: bool = False) -> "Repository":
+            and dependency.constraint.allows(package.version)
+        ]
+
+        # If we've previously made a choice that is compatible with the current
+        # requirement, stick with it.
+        for package in packages:
+            old_decision = decided.get(package)
+            if (
+                old_decision is not None
+                and not old_decision.marker.intersect(dependency.marker).is_empty()
+            ):
+                return package
+
+        return next(iter(packages), None)
 
     @classmethod
-    def __walk_dependency_level(
+    def __walk_dependencies(
         cls,
         dependencies: List[Dependency],
-        level: int,
-        pinned_versions: bool,
         packages_by_name: Dict[str, List[Package]],
-        project_level_dependencies: Set[str],
-        nested_dependencies: Dict[Tuple[str, str], Dependency],
-    ) -> Dict[Tuple[str, str], Dependency]:
-        if not dependencies:
-            return nested_dependencies
-
-        next_level_dependencies = []
-
-        for requirement in dependencies:
-            key = (requirement.name, requirement.pretty_constraint)
-            locked_package = cls.__get_locked_package(requirement, packages_by_name)
-
-            if locked_package:
-                # create dependency from locked package to retain dependency metadata
-                # if this is not done, we can end-up with incorrect nested dependencies
-                constraint = requirement.constraint
-                pretty_constraint = requirement.pretty_constraint
-                marker = requirement.marker
-                requirement = locked_package.to_dependency()
-                requirement.marker = requirement.marker.intersect(marker)
+    ) -> Dict[Package, Dependency]:
+        nested_dependencies: Dict[Package, Dependency] = {}
 
-                key = (requirement.name, pretty_constraint)
-
-                if not pinned_versions:
-                    requirement.set_constraint(constraint)
+        visited: Set[Tuple[Dependency, "BaseMarker"]] = set()
+        while dependencies:
+            requirement = dependencies.pop(0)
+            if (requirement, requirement.marker) in visited:
+                continue
+            visited.add((requirement, requirement.marker))
 
-                for require in locked_package.requires:
-                    if require.marker.is_empty():
-                        require.marker = requirement.marker
-                    else:
-                        require.marker = require.marker.intersect(requirement.marker)
+            locked_package = cls.__get_locked_package(
+                requirement, packages_by_name, nested_dependencies
+            )
 
-                    require.marker = require.marker.intersect(locked_package.marker)
+            if not locked_package:
+                # Should normally be able to satisfy all requirements, but this case is
```

Comment:
I think this really ought not to happen, or if it does it indicates that something has gone wrong.  

The reason that I have left this check here rather than raising some sort of exception is the non-dev version of [this test](https://github.com/python-poetry/poetry/blob/31657e7e5c47da758c14645035e59eb7813ac2f8/tests/utils/test_exporter.py#L1527) in which:

- we find `bar`, which has the `baz` extra set
- but when we encounter `baz` we find that it is dev-only

I wonder whether it is even possible for a lockfile to be constructed this way; or anyway whether it doesn't show that something has gone wrong already.  Seems like `baz` should be a non-dev dependency, if there's a non-dev path to wanting it.

So perhaps we should remove that testcase, and be more assertive here that we expect to find a `locked_package` meeting whatever requirement we have reached.

Reply:
tweaked the testcase so that `bar` is a dev dependency, which I think is likely what was intended

Reply:
changed my mind about this after encountering a real life example, reinstated the original testcase and made a fix to the code to pay more attention to what extras we want